README.md
AzaPhpGen
=========

Anizoptera CMF PHP code generation (dump, serialization) component.

https://github.com/Anizoptera/AzaPhpGen

[![Build Status][TravisImage]][Travis]


Table of Contents
-----------------

1. [Introduction](#introduction)
2. [Requirements](#requirements)
3. [Installation](#installation)
4. [Examples](#examples)
   * [Simple dump](#example-1---simple-dump)
   * [Array dump](#example-2---array-dump)
   * [Traversable dump](#example-3---traversable-dump)
   * [Closure (anonymous function) example](#example-4---closure-anonymous-function-example)
   * [Custom object dumping with IPhpGenerable interface](#example-5---custom-object-dumping-with-iphpgenerable-interface)
   * [Bundled CustomCode class usage](#example-6---bundled-customcode-class-usage)
   * [Custom object dumping with defined handlers](#example-7---custom-object-dumping-with-defined-handlers)
   * [AzaPhpGen customization](#example-8---azaphpgen-customization)
5. [Tests](#tests)
6. [Credits](#credits)
7. [License](#license)
8. [Links](#links)


Introduction
------------

Allows to dump complex arrays, objects, closures and basic data types as php code.
In part, this can be called a some sort of serialization.
And you can customize your dumped php code as you wish.

It is very usefull for code compilation (usually for caching purposes).


**Features:**

- Supports all scalar values (bool, int, float, string), nulls, arrays, serializable objects;
- [Traversable](http://php.net/traversable) support (dumped as array, see usage in [Example #3](#example-3---traversable-dump));
- Closures support (closures with "use", several closures on the same line are not supported!) (see usage and more info in [Example #4](#example-4---closure-anonymous-function-example));
- Custom object dumping with [IPhpGenerable interface](IPhpGenerable.php) (see usage in [Example #5](#example-5---custom-object-dumping-with-iphpgenerable-interface));
- Bundled simple [CustomCode class](CustomCode.php) (see usage in [Example #6](#example-6---bundled-customcode-class-usage));
- Custom object dumping with defined handlers/hooks (see usage in [Example #7](#example-7---custom-object-dumping-with-defined-handlers));
- Very flexible configuration (9 code building options, see in [PhpGen class code](PhpGen.php#L19));
- Automatic recognition of binary strings;
- Convenient, fully documented and test covered API;


**Benefits over `var_export()`:**

- `var_export` does not support Closures dumping;
- `var_export` supports only objects with `__set_state` function. AzaPhpGen supports all serializable objects;
- AzaPhpGen dumps Traversable objects as arrays (via `iterator_to_array`);
- For binary strings `var_export` generates very ugly code that is awkward to use and can be easily corrupted;
- For objects `var_export` generates code that can not be evaluated in namespace;
- AzaPhpGen give you full control over objects dumping with custom handlers and `IPhpGenerable` interface;
- With AzaPhpGen you can flexibly customize formatting of your code (useful for arrays);
- AzaPhpGen can generate code with or without trailing semicolon. `var_export` never outputs it :)
- Some detailed comparisons you can see in [Tests/PhpGenBenchmarkTest.php](Tests/PhpGenBenchmarkTest.php#L647);


Requirements
------------

* PHP 5.3.3 (or later);
* SPL and Reflection extensions for closures support (both bundled with PHP by default);


Installation
------------

The recommended way to install AzaPhpGen is [through composer](http://getcomposer.org).
You can see [package information on Packagist][ComposerPackage].

```JSON
{
	"require": {
		"aza/phpgen": "~1.0"
	}
}
```


Examples
--------

You can use [examples/example.php](examples/example.php) to run all examples.

#### Example #1 - Simple dump

```php
// Get singleton instance of PhpGen (fast and simple variant)
$phpGen = PhpGen::instance();
// Integer
echo $phpGen->getCode(123456789) . PHP_EOL; // 123456789;
// String (binary strings are supported as well)
echo $phpGen->getCode('some string' . ' example') . PHP_EOL; // "some string example";
// Float without trailing semicolon
echo $phpGen->getCodeNoTail(12.345) . PHP_EOL; // 12.345
// Simple serializable objects
$var = new stdClass();
echo $phpGen->getCode($var) . PHP_EOL; // unserialize("O:8:\"stdClass\":0:{}");
// Another object example
$var = new DateTime('2013-02-23 00:49:36', new DateTimeZone('UTC'));
echo $phpGen->getCode($var) . PHP_EOL; // unserialize("O:8:\"DateTime\":3:{s:4:\"date\";s:19:\"2013-02-23 00:49:36\";s:13:\"timezone_type\";i:3;s:8:\"timezone\";s:3:\"UTC\";}");
```

#### Example #2 - Array dump

```php
// AzaPhpGen will use short array syntax if possible by default (PHP >= 5.4)
echo $phpGen->getCode(array(
	true, false, null
)) . PHP_EOL;
/*
[
	true,
	false,
	null,
];
 */

// Build code without formatting
echo $phpGen->getCodeNoFormat(array(
	true, false, null
)) . PHP_EOL;
/*
[true,false,null];
 */

// Complex array (some sort of config for example)
$array = array(
	'key1'     => 'value',
	'long_key' => 'value',
	'array'    => array(
		'short_value'
	),
	'array2' => array(
		'very very very very very very very very very very very very long value'
	),
	'other',
	123456789
);
echo $phpGen->getCode($array) . PHP_EOL;
/*
[
	"key1"     => "value",
	"long_key" => "value",
	"array"    => ["short_value"],
	"array2"   => [
		"very very very very very very very very very very very very long value",
	],
	0 => "other",
	1 => 123456789,
];
 */

// And wothout formatting
echo $phpGen->getCodeNoFormat($array) . PHP_EOL;
/*
["key1"=>"value","long_key"=>"value","array"=>["short_value"],"array2"=>["very very very very very very very very very very very very long value"],0=>"other",1=>123456789];
 */
```

#### Example #3 - Traversable dump

AzaPhpGen treat all Traversable objects as arrays (with [iterator_to_array](http://php.net/iterator-to-array)).

```php
$var = new SplFixedArray(3);
$var[0] = 'a';
$var[1] = 'b';
echo $phpGen->getCodeNoFormat($var) . PHP_EOL; // ["a","b",null];
```

#### Example #4 - Closure (anonymous function) example

**WARNING:** Closures are dumped as is. So complex closures are not supported:
* Closures with "use" statement (closures that inherit variables from the parent scope);
* Several closures on the same line;
* Usage of non-qualified class name (with importing) in closure;
* Closures with `$this` variable usage;

```php
$closure = function($a, $b) {
	return round($a, $b) . "example\t\n";
};
echo $phpGen->getCode($closure) . PHP_EOL;
/*
function($a, $b) {
	return round($a, $b) . "example\t\n";
};
 */
echo $phpGen->getCode(array('key' => $closure)) . PHP_EOL;
/*
[
	"key" => function($a, $b) {
	return round($a, $b) . "example\t\n";
},
];
 */
```

#### Example #5 - Custom object dumping with IPhpGenerable interface

You can customize dumping of your classes by implementing the `IPhpGenerable` interface.

```php
class ExampleCustomCode implements IPhpGenerable
{
	public function generateCode()
	{
		return '32434 + 5678';
	}
}

$var = new ExampleCustomCode();

echo $phpGen->getCode($var) . PHP_EOL; // 32434 + 5678;

echo $phpGen->getCode(array($var)) . PHP_EOL; // [32434 + 5678];
```

#### Example #6 - Bundled CustomCode class usage

For the simpliest varint of `IPhpGenerable` interface usages you can use bundled class - `CustomCode`.
It just takes the required code as a constructor argument.

```php
$var = new CustomCode('"some code" . PHP_EOL');

echo $phpGen->getCode($var) . PHP_EOL; // "some code" . PHP_EOL;

echo $phpGen->getCode(array($var)) . PHP_EOL; // ["some code" . PHP_EOL];
```

#### Example #7 - Custom object dumping with defined handlers

Second varint of resulting code customization - usage of defined handlers (hooks) for the classes.
This way you can customize dump of any possible class!

```php
// Set custom handler for DateTime type
$phpGen->addCustomHandler('DateTime', function($data) use ($phpGen) {
	/** @var $data \DateTime */
	return $phpGen->getCodeNoTail(
		$data->format("Y-m-dO")
	);
});
// Build code
$var = new DateTime('2013-02-23 00:49:36', new DateTimeZone('UTC'));
echo $phpGen->getCode($var) . PHP_EOL; // "2013-02-23+0000";
```

#### Example #8 - AzaPhpGen customization

AzaPhpGen has many options. So it's very simple to configure your resulting code
for your special needs (code style for example).
You can see all available options in the [PhpGen class code](PhpGen.php#L19).

```php
// Disable short array syntax and use 6 spaces for indentation
$phpGen->shortArraySyntax = false;
$phpGen->useSpaces        = true;
$phpGen->tabLength        = 6;
$var = array(array(array(23 => 'example')));
echo $phpGen->getCode($var) . PHP_EOL;
/*
array(
      array(
            array(
                  23 => "example",
            ),
      ),
);
 */
```


Tests
-----

Tests are in the `Tests` folder and reach 100% code-coverage.
To run them, you need PHPUnit.
Example:

    $ phpunit --configuration phpunit.xml.dist

Or with coverage report:

    $ phpunit --configuration phpunit.xml.dist --coverage-html code_coverage/


Credits
-------

AzaPhpGen is a part of [Anizoptera CMF][], written by [Amal Samally][] (amal.samally at gmail.com) and [AzaGroup][] team.


License
-------

Released under the [MIT](LICENSE.md) license.


Links
-----

* [Composer package][ComposerPackage]
* [Last build on the Travis CI][Travis]
* [Project profile on the Ohloh](https://www.ohloh.net/p/AzaPhpGen)
* Other Anizoptera CMF components on the [GitHub][Anizoptera CMF] / [Packagist](https://packagist.org/packages/aza)
* (RU) [AzaGroup team blog][AzaGroup]



[Anizoptera CMF]:  https://github.com/Anizoptera
[Amal Samally]:    http://azagroup.ru/about/#amal
[AzaGroup]:        http://azagroup.ru/
[ComposerPackage]: https://packagist.org/packages/aza/phpgen
[TravisImage]:     https://secure.travis-ci.org/Anizoptera/AzaPhpGen.png?branch=master
[Travis]:          http://travis-ci.org/Anizoptera/AzaPhpGen


CHANGELOG.md
CHANGELOG
=========

## Version 1.2 (work in progress)


## Version 1.1.1 (28.05.2013)
07.04.2013
- **MINOR:** Massive README and CHANGELOG improvements (amal)


## Version 1.1 (25.03.2013)
- **MINOR:** Detailed comparision with `var_export` added (Issue #1, amal)
- **MINOR:** Test for resource type (amal)
- **FEATURE:** Automatic recognition of binary strings (amal)
- **IMPROVED:** Binary strings output improvement (encode '"', '\\', '$') (amal)
- **MINOR:** More tests for resulting code (NAN, INF) (amal)


## Version 1.0 (26.02.2013)
- First public release


PhpGen.php
<?php

namespace Aza\Components\PhpGen;
use ReflectionFunction;
use SplFileObject;
use Traversable;

/**
 * PHP code generation
 *
 * @uses reflection
 * @uses spl
 *
 * @project Anizoptera CMF
 * @package system.phpgen
 * @author  Amal Samally <amal.samally at gmail.com>
 * @license MIT
 */
class PhpGen
{
	/**
	 * Length of one tab in spaces
	 */
	public $tabLength = 4;

	/**
	 * Use spaces instead of tabs for indentation
	 */
	public $useSpaces = false;

	/**
	 * Mix tabs with spaces in the end of indentation
	 */
	public $mixSpaces = true;

	/**
	 * Mix tabs with spaces in the end of indentation
	 */
	public $spacesAfterKey = true;

	/**
	 * Output string variables as one line.
	 * Converts \n and \t symbols to escaped characters.
	 *
	 * By default only other control and non-visible
	 * chars are escaped.
	 */
	public $oneLineStrings = false;

	/**
	 * Escape all characters in strings.
	 */
	public $binaryStrings = false;

	/**
	 * Automatic recognition of binary strings
	 */
	public $binaryAutoCheck = true;

	/**
	 * Output array keys for serial arrays.
	 */
	public $outputSerialKeys = false;

	/**
	 * Use php 5.4 short array syntax
	 */
	public $shortArraySyntax = false;

	/**
	 * If array value breaks the list (multiline),
	 * alignment before and after calculated separately
	 */
	public $alignMultilineBreaks = true;

	/**
	 * Approximate maximum line length
	 */
	public $maxLineLength = 60;


	/**
	 * Array of different custom type handlers.
	 *
	 * @var array[]
	 */
	protected $customHandlers = array();



	/**
	 * Returns singleton instance of the class
	 *
	 * @return self
	 */
	public static function instance()
	{
		static $instance;
		return $instance ?: $instance = new self;
	}

	/**
	 * Constructor
	 */
	public function __construct()
	{
		$this->shortArraySyntax = version_compare(PHP_VERSION, '5.4', '>=');
	}


	/**
	 * Adds custom handler for the specified object type
	 *
	 * @param string|object $type <p>
	 * Object type for instanceof check
	 * </p>
	 * @param callable $handler <p>
	 * Callback "string fun($object)"
	 * </p>
	 */
	public function addCustomHandler($type, $handler)
	{
		$this->customHandlers[] = array($type, $handler);
	}


	/**
	 * Generation of php code for various data without formatting.
	 *
	 * WARNING!
	 * Don't use for self referencing arrays and objects!
	 *
	 * @see getCode
	 *
	 * @param mixed $data   Data
	 * @param bool  $noTail Don't add trailing ";"
	 *
	 * @return string
	 */
	public function getCodeNoFormat($data, $noTail = false)
	{
		return $this->getCode($data, 0, true, $noTail);
	}

	/**
	 * Generation of php code for various data without trailing semicolon (;).
	 *
	 * WARNING!
	 * Don't use for self referencing arrays and objects!
	 *
	 * @see getCode
	 *
	 * @param mixed $data     Data
	 * @param int   $indent   Indent size in tabs for array php code
	 * @param bool  $noFormat No formatting and indention
	 *
	 * @return string
	 */
	public function getCodeNoTail($data, $indent = 0, $noFormat = false)
	{
		return $this->getCode($data, $indent, $noFormat, true);
	}

	/**
	 * Generation of php code for various data.
	 *
	 * WARNING!
	 * Don't use for self referencing arrays and objects!
	 *
	 * @param mixed $data     Data
	 * @param int   $indent   Indent size in tabs for array php code
	 * @param bool  $noFormat No formatting and indention
	 * @param bool  $noTail   Don't add trailing semicolon (;)
	 *
	 * @return string
	 */
	public function getCode($data, $indent = 0, $noFormat = false, $noTail = false)
	{
		$tail = $noTail ? '' : ';';

		// Null
		if (!isset($data)) {
			// var_export returns uppercased, so use own variant
			return 'null' . $tail;
		}
		// Bool / Int / Float
		else if (is_bool($data) || is_int($data) || is_float($data)) {
			return var_export($data, true) . $tail;
		}
		// Array
		else if (($traversable = ($data instanceof Traversable)) || is_array($data)) {
			if ($traversable) {
				$data = iterator_to_array($data, true);
			}
			return $this->getArray($data, $indent, $noFormat) . $tail;
		}
		// Object
		else if (is_object($data)) {
			return $this->getObject($data, $indent, $noFormat) . $tail;
		}
		// String
		return $this->getString((string)$data) . $tail;
	}


	/**
	 * Returns php code for string
	 *
	 * @link http://php.net/language.types.string#language.types.string.syntax.double
	 *
	 * @param string $string <p>
	 * String data
	 * </p>
	 *
	 * @return string
	 */
	protected function getString($string)
	{
		// Binary strings are completely encoded in contrast to ordinary
		// So we try to distinguish between a binary string and a usual
		if (!($binary = $this->binaryStrings) && $this->binaryAutoCheck) {
			// Simple check based on PCRE error for malformed UTF-8 data (not work for valid ASCII)
			$binary = false === preg_match('~~u', $string)
			       // Match for most non printable chars somewhat taking multibyte chars into account
			       // Variant from sebastianbergmann/exporter by Sebastian Bergmann <sebastian@phpunit.de>
			       // Works good for binary strings that consist of valid ASCII chars
			       || preg_match('~[^\x09-\x0d\x20-\xff]~S', $string);
		}

		// Build regexp
		if ($binary) {
			$regexp = '~.~Ss';
		} else {
			$regexp = $this->oneLineStrings
				// All control chars
					? '\x00-\x1F\x7F'
				// Ctrl chars without \n & \t
					: '\x00-\x08\x0B-\x1F\x7F';
			// Additional chars: \x22 ("), \x24 ($), \x5C (\\)
			$regexp = '~['.$regexp.'\x22\x24\x5C]~S';
		}

		// Format string
		$string = preg_replace_callback(
			$regexp,
			function($char) use ($binary) {
				// linefeed (LF or 0x0A (10) in ASCII)
				if ("\n" === ($char = $char[0])) {
					return '\n';
				}
				// carriage return (CR or 0x0D (13) in ASCII)
				else if ("\r" === $char) {
					return '\r';
				}
				// horizontal tab (HT or 0x09 (9) in ASCII)
				else if ("\t" === $char) {
					return '\t';
				}
				// vertical tab (VT or 0x0B (11) in ASCII) (since PHP 5.2.5)
				else if ("\v" === $char) {
					return '\v';
				}
				// escape (ESC or 0x1B (27) in ASCII) (since PHP 5.4.0)
				// works only for PHP 5.4.0 or above, so disable
//				else if ("\e" === $char) {
//					return '\e';
//				}
				// form feed (FF or 0x0C (12) in ASCII) (since PHP 5.2.5)
				else if ("\f" === $char) {
					return '\f';
				}
				// chars that must be escaped in a double-quoted string
				else if (!$binary && ('\\' === $char || '$' === $char || '"' === $char)) {
					return "\\$char";
				}
				// all other chars
				return sprintf('\x%02X', ord($char));
			},
			(string)$string
		);

		return '"' . $string . '"';
	}

	/**
	 * Returns php code for object
	 *
	 * @see IPhpGenerable
	 * @see CustomCode
	 *
	 * @param object|IPhpGenerable $object <p>
	 * Object data
	 * </p>
	 *
	 * @return string
	 */
	protected function getObject($object)
	{
		// User custom code
		if ($object instanceof IPhpGenerable) {
			return $object->generateCode();
		}
		// Closures special (partial) support
		// WARNING: many closures on one line are not supported
		// WARNING: closures with "use" are not supported
		else if ($object instanceof \Closure) {
			$ref = new ReflectionFunction($object);

			// Open file and seek to the first line of the closure
			$file = new SplFileObject($ref->getFileName());
			$file->seek($ref->getStartLine()-1);

			// Retrieve all of the lines that contain code for the closure
			$endLine = $ref->getEndLine();
			$code    = '';
			while ($file->key() < $endLine) {
				$code .= $file->current();
				$file->next();
			}

			// Only keep the code defining that closure
			$begin = stripos($code, 'function');
			$end   = strrpos($code, '}');
			$code  = substr($code, $begin, $end - $begin + 1);

			return $code;
		}
		// Different custom handlers
		else if ($handlers = $this->customHandlers) {
			foreach ($handlers as $h) {
				list($type, $handler) = $h;
				if ($object instanceof $type) {
					return $handler($object);
				}
			}
		}
		// TODO: Add __set_state function support?
		// Default - serialization
		return "unserialize({$this->getCodeNoTail(serialize($object))})";
	}

	/**
	 * Returns php code for array
	 *
	 * @param array|Traversable $array <p>
	 * Array data
	 * </p>
	 * @param int $indent [optional] <p>
	 * Indent size in tabs for array php code
	 * </p>
	 * @param bool $noFormat [optional] <p>
	 * No formatting and indention
	 * </p>
	 *
	 * @return string
	 */
	protected function getArray($array, $indent = 0, $noFormat = false)
	{
		$shortSyntax = $this->shortArraySyntax;
		$resultCode  = $shortSyntax ? '[' : 'array(';

		if ($array) {
			$newLine      = "\n";
			$tabLength    = $this->tabLength;
			$mixSpaces    = $this->mixSpaces;
			$useSpaces    = $this->useSpaces;
			$tab          = $useSpaces ? str_repeat(' ', $tabLength) : "\t";
			$spacePostfix = $useSpaces || $this->spacesAfterKey;
			$alignBreaks  = $this->alignMultilineBreaks;

			// The overall indent
			$indentString = $noFormat ? '' : str_repeat($tab, $indent);

			// First calculations and code build for keys/values
			$keyLength = $keyLengthBlock = $i = 0;
			$maxKeyLenBlocks = array($keyLengthBlock => 0);
			$maxKeyLength    = &$maxKeyLenBlocks[$keyLengthBlock];
			$arrayIsSimple   = !$this->outputSerialKeys;
			$multiline       = false;
			$arrayParts      = array();
			foreach ($array as $key => $val) {
				if ($arrayIsSimple && $key !== $i++) {
					$arrayIsSimple = false;
				}

				// Build code for keys and values
				$key = $this->getCode($key, 0, true, true);
				$val = $this->getCode($val, $indent+1, $noFormat, true);

				// We don't need this information if formatting is disabled
				if (!$noFormat) {
					// We need to save maximum key code length for alignment
					$keyLength = mb_strlen($key, 'UTF-8');
					$keyLength > $maxKeyLength
						&& $maxKeyLength = $keyLength;

					// Multiline value breaks the list, so
					// alignment before and after calculated separately
					if ($multiline = $alignBreaks
							? false !== strpos($val, $newLine)
							: false
					) {
						$maxKeyLenBlocks[++$keyLengthBlock] = 0;
						$maxKeyLength = &$maxKeyLenBlocks[$keyLengthBlock];
					}
				}

				$arrayParts[] = array(
					$key,
					$val,
					$keyLength,
					$multiline
				);
			}
			unset($array, $maxKeyLength);

			// Disable formatting if array have only one serial value
			// (not multiline and not too long)
			if ($arrayIsSimple && !$noFormat && count($arrayParts) === 1
			    && !$arrayParts[0][3]
			    && $this->maxLineLength >= mb_strlen($arrayParts[0][1], 'UTF-8')
			                               + ($indent*$tabLength)
			) {
				$noFormat = true;
			}

			// Build code
			$keyLengthBlock = 0;
			$maxKeyLength   = $maxKeyLenBlocks[$keyLengthBlock];
			foreach ($arrayParts as &$data) {
				list(
					$key,
					$val,
					$keyLength,
					$multiline
				) = $data;

				if (!$noFormat) {
					// Full align length in chars (after key, before value)
					$alignLength = $maxKeyLength - $keyLength + 1;

					// Simple spaces alignment
					if ($spacePostfix) {
						$key .= str_repeat(' ', $alignLength);
					}

					// Tabs with mix of spaces alignment (for shortness)
					else if ($mixSpaces) {
						if ($curTabTail = $keyLength % $tabLength) {
							$curTabTail = $tabLength-$curTabTail;
						}
						$alignTail = $alignLength % $tabLength;
						if ($curTabTail <= $alignTail
							|| $alignLength / $tabLength >= 1
						) {
							if ($curTabTail) {
								$key         .= $tab;
								$alignLength -= $curTabTail;
							}
							if ($alignLength) {
								if ($align = floor($alignLength / $tabLength)) {
									$key .= str_repeat($tab, $align);
								}
								$alignTail = $alignLength % $tabLength;
							} else {
								$alignTail = 0;
							}
						}
						if ($alignTail) {
							$key .= str_repeat(' ', $alignTail);
						}
					}

					// Only tabs alignment
					else {
						if ($curTabTail = $keyLength % $tabLength) {
							$key         .= $tab;
							$alignLength -= $tabLength-$curTabTail;
						}
						$key .= str_repeat(
							$tab,
							ceil($alignLength / $tabLength)
						);
					}

					// Different align before and after multiline value
					if ($multiline) {
						$maxKeyLength = $maxKeyLenBlocks[++$keyLengthBlock];
					}
				}

				$data = ($arrayIsSimple
							? ''
							: $key . '=>' . ($noFormat ? '' : ' ')
				        ) . $val . ',';
			}

			// Join code parts
			if ($noFormat) {
				$resultCode .= join('', $arrayParts);
				$resultCode  = substr($resultCode, 0, -1);
			} else {
				$resultCode .= "{$newLine}{$indentString}{$tab}"
						. join("{$newLine}{$indentString}{$tab}", $arrayParts)
						. "{$newLine}{$indentString}";
			}
		}

		$resultCode .= $shortSyntax ? ']' : ')';

		return $resultCode;
	}
}


examples/example.php
<?php

use Aza\Components\PhpGen\CustomCode;
use Aza\Components\PhpGen\IPhpGenerable;
use Aza\Components\PhpGen\PhpGen;

require __DIR__ . '/../vendor/autoload.php';


/**
 * AzaPhpGen examples
 *
 * @project Anizoptera CMF
 * @package system.phpgen
 * @author  Amal Samally <amal.samally at gmail.com>
 * @license MIT
 */



// Example #1 - Simple dump
// ------------------------

// Get singleton instance of PhpGen
$phpGen = PhpGen::instance();
// Integer
echo $phpGen->getCode(123456789) . PHP_EOL; // 123456789;
// String (binary strings are supported as well)
echo $phpGen->getCode('some string' . ' example') . PHP_EOL; // "some string example";
// Float without trailing semicolon
echo $phpGen->getCodeNoTail(12.345) . PHP_EOL; // 12.345
// Simple serializable objects
$var = new stdClass();
echo $phpGen->getCode($var) . PHP_EOL; // unserialize("O:8:\"stdClass\":0:{}");
// Another object example
$var = new DateTime('2013-02-23 00:49:36', new DateTimeZone('UTC'));
echo $phpGen->getCode($var) . PHP_EOL; // unserialize("O:8:\"DateTime\":3:{s:4:\"date\";s:19:\"2013-02-23 00:49:36\";s:13:\"timezone_type\";i:3;s:8:\"timezone\";s:3:\"UTC\";}");


// Example #2 - Array dump
// ------------------------

// AzaPhpGen will use short array syntax if possible by default (PHP >= 5.4)
echo $phpGen->getCode(array(
	true, false, null
)) . PHP_EOL;
/*
[
	true,
	false,
	null,
];
 */

// Build code without formatting
echo $phpGen->getCodeNoFormat(array(
	true, false, null
)) . PHP_EOL;
/*
[true,false,null];
 */

// Complex array (some sort of config for example)
$array = array(
	'key1'     => 'value',
	'long_key' => 'value',
	'array'    => array(
		'short_value'
	),
	'array2' => array(
		'very very very very very very very very very very very very long value'
	),
	'other',
	123456789
);
echo $phpGen->getCode($array) . PHP_EOL;
/*
[
	"key1"     => "value",
	"long_key" => "value",
	"array"    => ["short_value"],
	"array2"   => [
		"very very very very very very very very very very very very long value",
	],
	0 => "other",
	1 => 123456789,
];
 */

// And wothout formatting
echo $phpGen->getCodeNoFormat($array) . PHP_EOL;
/*
["key1"=>"value","long_key"=>"value","array"=>["short_value"],"array2"=>["very very very very very very very very very very very very long value"],0=>"other",1=>123456789];
 */



// Example #3 - Traversable dump
// ------------------------

$var = new SplFixedArray(3);
$var[0] = 'a';
$var[1] = 'b';
echo $phpGen->getCodeNoFormat($var) . PHP_EOL; // ["a","b",null];



// Example #4 - Closure example
// ------------------------

$closure = function($a, $b) {
	return round($a, $b) . "example\t\n";
};
echo $phpGen->getCode($closure) . PHP_EOL;
/*
function($a, $b) {
	return round($a, $b) . "example\t\n";
};
 */
echo $phpGen->getCode(array('key' => $closure)) . PHP_EOL;
/*
[
	"key" => function($a, $b) {
	return round($a, $b) . "example\t\n";
},
];
 */



// Example #5 - Custom object dumping with IPhpGenerable interface
// ------------------------

class ExampleCustomCode implements IPhpGenerable
{
	public function generateCode()
	{
		return '32434 + 5678';
	}
}

$var = new ExampleCustomCode();

echo $phpGen->getCode($var) . PHP_EOL; // 32434 + 5678;

echo $phpGen->getCode(array($var)) . PHP_EOL; // [32434 + 5678];



// Example #6 - Bundled simple CustomCode class usage
// ------------------------

$var = new CustomCode('"some code" . PHP_EOL');

echo $phpGen->getCode($var) . PHP_EOL; // "some code" . PHP_EOL;

echo $phpGen->getCode(array($var)) . PHP_EOL; // ["some code" . PHP_EOL];



// Example #7 - Custom object dumping with defined handlers
// ------------------------

// Set custom handler for DateTime type
$phpGen->addCustomHandler('DateTime', function($data) use ($phpGen) {
	/** @var $data \DateTime */
	return $phpGen->getCodeNoTail(
		$data->format("Y-m-dO")
	);
});
// Build code
$var = new DateTime('2013-02-23 00:49:36', new DateTimeZone('UTC'));
echo $phpGen->getCode($var) . PHP_EOL; // "2013-02-23+0000";



// Example #8 - AzaPhpGen customization
// ------------------------

// Disable short array syntax and use 6 spaces for indentation
$phpGen->shortArraySyntax = false;
$phpGen->useSpaces        = true;
$phpGen->tabLength        = 6;
$var = array(array(array(23 => 'example')));
echo $phpGen->getCode($var) . PHP_EOL;
/*
array(
      array(
            array(
                  23 => "example",
            ),
      ),
);
 */
